### PR TITLE
[3.16] backport: #11009 #10962

### PR DESCRIPTION
**Pull request by maiste**
This PR backports https://github.com/ocaml/dune/pull/11009 and #10962. They are backported together as one is the fix of the other one.

### CODE REVIEW

File: test/blackbox-tests/test-cases/cpp-std-ocaml5.t/cpp11.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include <caml/misc.h>
+#include <caml/mlvalues.h>
+#include <iostream>
+#include <cstring>
```
Comment (kit-ty-kate) [inline test/blackbox-tests/test-cases/cpp-std-ocaml5.t/cpp11.cpp:4]:
```suggestion
#include <iostream>
#include <cstring>
#include <caml/misc.h>
#include <caml/mlvalues.h>
```

File: src/ocaml/version.ml [hunk 2]
```diff
@@ -29,3 +29,5 @@ let supports_alerts version = version >= (4, 8, 0)
 let has_sandboxed_otherlibs version = version >= (5, 0, 0)
 let has_META_files version = version >= (5, 0, 0)
 let supports_bin_annot_occurrences version = version >= (5, 2, 0)
+let supports_hidden_includes version = version >= (5, 2, 0)
```
Comment (Leonidas-from-XIV) [inline src/ocaml/version.ml:32]: This seems questionable, I don't think it is part of the fix.

File: src/ocaml/version.mli [hunk 3]
```diff
@@ -76,3 +76,8 @@ val has_META_files : t -> bool
 
 (** Whether the compiler supports occurrences indexation *)
 val supports_bin_annot_occurrences : t -> bool
+
+(** Whether the compiler supports the -H flag *)
+val supports_hidden_includes : t -> bool
```
Comment (Leonidas-from-XIV) [inline src/ocaml/version.mli:81]: Same here.

**Review by Leonidas-from-XIV — CHANGES_REQUESTED** [review]

File: src/ocaml/version.ml (hunk 2 above)
Comment (maiste) [inline src/ocaml/version.ml:32]:
```suggestion
```

File: src/ocaml/version.mli (hunk 3 above)
Comment (maiste) [inline src/ocaml/version.mli:81]:
```suggestion
```

**Review by maiste — COMMENTED** [review]
Let see if the tests succeed. If that's the case, I'll commit these changes.

**Review by Leonidas-from-XIV — APPROVED** [review]

**Pull request merged**